test/functional/tests/data_integrity/test_data_integrity_5d.py

```diff
@@ -0,0 +1,214 @@
+        pass_criteria:
+            - System does not crash.
+            - All operations complete successfully.
+            - Data consistency is being preserved.
+    """
+    with TestRun.step("Prepare cache and core devices"):
+        cache_devices, core_devices = prepare_devices()
+
+    with TestRun.step("Run 4 cache instances in different cache modes, add single core to each"):
+        cache_modes = [CacheMode.WT, CacheMode.WB, CacheMode.WA, CacheMode.WO]
+        caches = []
+        cores = []
+        for i in range(4):
+            cache, core = start_instance(cache_devices[i], core_devices[i], cache_modes[i])
+            caches.append(cache)
+            cores.append(core)
+
+    with TestRun.step("Load default io class config for each cache"):
+        for cache in caches:
+            cache.load_io_class("/etc/opencas/ioclass-config.csv")
+
+    with TestRun.step("Create filesystems and mount cores"):
+        for i, core in enumerate(cores):
+            mount_point = core.system_path.replace('/dev/', '/mnt/')
+            if not fs_utils.check_if_directory_exists(mount_point):
+                fs_utils.create_directory(mount_point)
+            TestRun.LOGGER.info(f"Create filesystem {filesystems[i].name} on {core.system_path}")
+            core.create_filesystem(filesystems[i])
+            TestRun.LOGGER.info(f"Mount filesystem {filesystems[i].name} on {core.system_path} to "
+                                f"{mount_point}")
+            core.mount(mount_point)
+            sync()
+
+    with TestRun.step("Run test workloads on filesystems with verification"):
+        fio_run = Fio().create_command()
+        fio_run.io_engine(IoEngine.libaio)
+        fio_run.direct()
+        fio_run.time_based()
+        fio_run.nr_files(4096)
+        fio_run.file_size_range([(file_min_size, file_max_size)])
+        fio_run.do_verify()
+        fio_run.verify(VerifyMethod.md5)
+        fio_run.verify_dump()
+        fio_run.run_time(runtime)
+        fio_run.read_write(ReadWrite.randrw)
+        fio_run.io_depth(128)
+        fio_run.blocksize_range([(start_size, stop_size)])
+        for core in cores:
+            fio_job = fio_run.add_job()
+            fio_job.directory(core.mount_point)
+            fio_job.size(core.size)
+        fio_run.run()
+
+    with TestRun.step("Unmount cores"):
+        for core in cores:
+            core.unmount()
+
+    with TestRun.step("Calculate md5 for each core"):
+        core_md5s = [File(core.full_path).md5sum() for core in cores]
+
+    with TestRun.step("Stop caches"):
+        for cache in caches:
+            cache.stop()
+
+    with TestRun.step("Calculate md5 for each core"):
+        dev_md5s = [File(dev.full_path).md5sum() for dev in core_devices]
+
+    with TestRun.step("Compare md5 sums for cores and core devices"):
+        for i in core_md5, dev_md5, mode, fs in zip(core_md5s, dev_md5s, cache_modes, filesystems):
```

Comment:
for core_md5, dev_md5, mode, fs in zip(core_md5s, dev_md5s, cache_modes, filesystems):